meson.build

```
@@ -0,0 +1,250 @@
+  'WINDOWS_VERSIONINFO': windows_versioninfo,
+}
+
+# Compiler flags
+cc = meson.get_compiler('c')
+cc_native = meson.get_compiler('c', native : true)
+add_project_arguments(
+  cc.get_supported_arguments(
+    '-Wstrict-prototypes',
+    '-Wmissing-prototypes',
+    '-Wmissing-declarations',
+    '-Wnested-externs',
+    '-fno-common',
+  ),
+  '-DG_DISABLE_SINGLE_INCLUDES',
+  '-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_56',
+  '-DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_MIN_REQUIRED',
+  language : 'c'
+)
+add_project_link_arguments(
+  cc.get_supported_link_arguments(
+    '-Wl,--no-undefined',
+  ),
+  language : 'c'
+)
+
+# fopen cloexec flag
+if host_machine.system() == 'windows'
+  message('Using "N" flag for close-on-exec (Windows)')
+  conf.set_quoted('FOPEN_CLOEXEC_FLAG', 'N')
+else
+  code = '''
+  #include <stdio.h>
+  #include <unistd.h>
+  #include <fcntl.h>
+
+  int main(int argc, char **argv) {
+    FILE *fp = fopen("/dev/null", "re");
+    if (fp != NULL) {
+      int fd = fileno(fp);
+      if (fd != -1) {
+        long ret = fcntl(fd, F_GETFD);
+        if (ret != -1 && (ret & FD_CLOEXEC)) {
+          return 0;
+        }
+      }
+    }
+    return 1;
+  }
+  '''
+  result = cc.run(code, name : 'check fopen() close-on-exec flag')
+  if result.compiled() and result.returncode() == 0
+    # glibc >= 2.7, FreeBSD >= 10.0, NetBSD >= 6.0
+    message('Using "e" flag for close-on-exec')
+    conf.set_quoted('FOPEN_CLOEXEC_FLAG', 'e')
+  else
+    message('Using no close-on-exec flag (unknown or cross compile)')
+    conf.set_quoted('FOPEN_CLOEXEC_FLAG', '')
+    conf.set('NONATOMIC_CLOEXEC', 1)
+  endif
+endif
+if get_option('_nonatomic_cloexec')
+  # CI sets this on Linux distros with libraries that don't correctly set
+  # CLOEXEC.  Disable CLOEXEC leak check.
+  conf.set('NONATOMIC_CLOEXEC', 1)
+endif
+
+# Math library, host and build
+foreach i : [[cc, 'libm_dep'], [cc_native, 'libm_native_dep']]
+  if not i[0].has_function('floor', prefix : '#include <math.h>')
+    set_variable(i[1], i[0].find_library('m'))
+  else
+    set_variable(i[1], declare_dependency())
+  endif
+endforeach
+
+# Dependencies
+feature_flags = []
+zlib_dep       = dependency('zlib')
+jpeg_dep       = dependency('libjpeg')
+png_dep        = dependency('libpng', version : '>1.2')
+openjpeg_dep   = dependency('libopenjp2', version : '>=2.1.0')
+tiff_dep       = dependency('libtiff-4')
+glib_dep       = dependency('glib-2.0', version : '>=2.56')
+gio_dep        = dependency('gio-2.0')
+gobject_dep    = dependency('gobject-2.0')
+cairo_dep      = dependency('cairo', version : '>=1.2')
+gdk_pixbuf_dep = dependency('gdk-pixbuf-2.0', version : '>=2.14')
+xml_dep        = dependency('libxml-2.0')
+sqlite_dep     = dependency('sqlite3', version : '>=3.14')
+valgrind_dep   = dependency('valgrind', required : false)
+
+doc = get_option('doc')
+if not doc.disabled()
+  doxygen = find_program('doxygen', required : doc)
+else
+  doxygen = disabler()
+endif
```

Hmm, can't you just use:
```suggestion
doxygen = find_program('doxygen', required : get_option('doc'), disabler: true)
```

---

As far as I understood the meson documentation, there is a subtle difference.
The comment for the disabler argument on find_program states

> If `true` and the program couldn't be found, return a [disabler](https://mesonbuild.com/Reference-manual_returned_disabler.html) object

So specifying the `disabler : true` argument means that documentation would still be built as long as doxygen is found, while the original if else solution disables the documentation build when the doc option is set to disabled.

It might still be useful to add disabler:true to the find_program line either way so that we disable documentation build when required is false.

---

Ah, reading more, it looks like you may be right and the trick is that when the 'required' argument is disable, it will not try to look for the doxygen binary and as a result fail to find it and return the disabler.

So yes, that change LGTM

---

Indeed, feature options with a disabled state are IMHO one of the coolest things Meson has, because it allows you to concisely express things that a boolean simply cannot -- and part of that is that feature options Do The Right Thing™ without needing bulky if/else constructs to open-code what "disabled" means.

It will even give you a pretty log message: "Program doxygen skipped: feature **doxygen** disabled".